### PR TITLE
Address #134 and #89: no http dependency for Endpoint

### PR DESCRIPTION
**Pull request by vkostyukov**
See details #134 and #89.

@pasviegas, @benjumanji, @rodrigopr, @rpless, @BenWhitehead please review.

### CODE REVIEW

File: finch-core/src/main/scala/io/finch/Endpoint.scala [hunk 1]
```diff
@@ -28,12 +28,13 @@ import com.twitter.finagle.httpx.path.Path
 import com.twitter.finagle.httpx.service.NotFoundService
 
 /**
- * A REST API endpoint that primary defines a ''route'' and might be converted
- * into a finagled service with ''toService'' method.
+ * A REST API endpoint that primary defines a ''route'' and might be
+ * implicitly converted into Finagle service ''Service[Req, Rep]''.
```
Comment (rodrigopr) [inline finch-core/src/main/scala/io/finch/Endpoint.scala]: I think it's better to describe that it requires an implicit `Route` to be converted to a Service

File: finch-core/src/main/scala/io/finch/package.scala [hunk 2]
```diff
@@ -147,11 +148,36 @@ package object finch {
...
+    new Service[Req, Rep] {
+      def apply(req: Req) = f
+    }
+
+  /**
+   * A function that fetches the route information from the given request ''req''.
+   *
+   * @tparam Req the request type
+   */
+  trait Route[-Req] {
```
Comment (rodrigopr) [inline finch-core/src/main/scala/io/finch/package.scala]: We could use `@implicitNotFound` annotation with further explanation which why the implicit is required.

Comment (vkostyukov) [inline finch-core/src/main/scala/io/finch/package.scala]: Nice! I didn't know that!

File: finch-core/src/main/scala/io/finch/Endpoint.scala (hunk 1 above)
Comment (vkostyukov) [inline finch-core/src/main/scala/io/finch/Endpoint.scala]: Sure!

**Comment by vkostyukov** [comment]
BTW, I was quite surprised how easy it is to make such compete and big changes in the Finch's code base. Not sure who to thank: Scala or Finch itself :)

**Comment by rodrigopr** [comment]
@vkostyukov, yeah hehe, also great that the api remained the same. nice job.

beside my last two comments it looks good to me.

File: finch-core/src/main/scala/io/finch/package.scala [hunk 3]
```diff
@@ -139,19 +141,42 @@ package object finch {
...
+   * Allows to use an ''Endpoint'' when ''Service'' is expected. The implicit
+   * conversion is possible if there is an implicit view from ''Req'' to
+   * ''HttpRequest'' available in the scope.
+   *
+   * {{{
+   *   val e: Endpoint[HttpRequest, HttpResponse] = ???
+   *   Httpx.serve(new InetSocketAddress(8081), e)
+   * }}}
+   *
+   * @param e the
```
Comment (rodrigopr) [inline finch-core/src/main/scala/io/finch/package.scala]: incomplete doc here

**Comment by vkostyukov** [comment]
Well, there is a new issue with `RequiredBody` and `OptionalBody` since they depends on two type parameters we can't really use them as previously: `RequiredBody[Json](req)`. I added two tests which are failed now. Hope I will come with something that will allow to keep the API unchanged. Travis CI should show the error message. For now it's:

```
error] /Users/vkostyukov/workspace/finch/finch-core/src/test/scala/io/finch/request/BodySpec.scala:101: type mismatch;
[error]  found   : io.finch.HttpRequest
[error]     (which expands to)  com.twitter.finagle.httpx.Request
[error]  required: io.finch.request.DecodeRequest[Int]
[error]     val i: Future[Int] = RequiredBody[Int](req)
[error]                                            ^
[error] /Users/vkostyukov/workspace/finch/finch-core/src/test/scala/io/finch/request/BodySpec.scala:101: type mismatch;
[error]  found   : io.finch.request.RequestReader[Int]
[error]  required: com.twitter.util.Future[Int]
[error]     val i: Future[Int] = RequiredBody[Int](req)
[error]                                           ^
```

@rodrigopr you have a better understanding of implicits then me. Do you have any idea what we can do here?

**Comment by vkostyukov** [comment]
There is a workaround: just use `RequiredBody(req)` instead of `RequiredBody[Json](req)`. I personally liked the original version much more since it indicates which type you're reading from the request.

**Comment by rodrigopr** [comment]
Yeah, I prefer the original version (`RequiredBody[Json](req)`) too.
Thinking about it, we could keep the code before the last change [1], as the user will have an implicit view `Req => HttpRequest` in scope anyway, it shouldn't be a problem.

[1] `def apply[A](req: HttpRequest)(implicit decode: DecodeRequest[A])`

**Comment by vkostyukov** [comment]
Right. I made the changes and added `TODO`. This will definitely work correct, since there is an implicit conversion available. It's just a bit confusing, but, for now it's a good solution.

**Comment by vkostyukov** [comment]
Last call @rpless, @benjumanji, @BenWhitehead, @pasviegas. Please review. I'm pretty happy with the state of this PR.

**Comment by BenWhitehead** [comment]
@vkostyukov Looks good to me, nice job on keeping the Api they while updating the type bounds.

**Comment by rpless** [comment]
:+1: for maintaining the API.

**Comment by vkostyukov** [comment]
There is just one thing I don't really like in the whole story: I added `TODO` for it. For now I don't really have a solution for that issue but I hope will come up with something later.

**Pull request merged**